OpenTelemetry output plugin

jacobmarble opened the pull request:
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

Related:
- Telegraf input plugin for OpenTelemetry https://github.com/influxdata/telegraf/pull/9077
- OpenTelemetry issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2951

This change adds an output plugin opentelemetry, which sends OTLP signals via gRPC.

[OpenTelemetry](https://opentelemetry.io/) is an open-source telemetry ecosystem, supporting traces, metrics and logs. The mapping of the OpenTelemetry data model to line protocol is a [work-in-progress](https://github.com/influxdata/influxdb-observability/blob/main/docs/index.md), and I welcome any feedback to that document.

The core conversion logic lives in influxdata/influxdb-observability/influx2otel, intended to stand as a canonical implementation of line protocol->OTel.

Note that, unlike the OpenTelemetry input plugin, this output plugin only supports metrics, and not traces or logs. That may be added later.

ssoroka commented on plugins/outputs/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,134 @@
+			o.Log.Warnf("unrecognized metric type %Q", metric.Type())
+			continue
+		}
+		err := batch.AddPoint(metric.Name(), metric.Tags(), metric.Fields(), metric.Time(), vType)
```
There's a panic in AddPoint that I think should just be an error. Telegraf really doesn't like panics. https://github.com/influxdata/influxdb-observability/blob/6ea7daf489f30bd9e7f7d3f7ef65e8074fb6a2ec/influx2otel/metrics_telegraf_prometheus_v1.go#L83

jacobmarble replied:
Fixed.

ssoroka reviewed (changes requested):
I'd like to distance this a bit from prometheus (otel has no direct relationship). Could also use tests.

jacobmarble commented:
@ssoroka friendly reminder

codeboten commented:
I ran a manual test using telegraf with this plugin to export data to an opentelemetry collector locally and didn't see any data being generated.

here's the output from the telegraf logs:
```
2021-05-12T21:12:35Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-05-12T21:12:35Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-05-12T21:12:35Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-05-12T21:12:35Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-05-12T21:12:35Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-05-12T21:12:35Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-05-12T21:12:35Z W! [outputs.opentelemetry] failed to add point: gauge field not found
```

and the data as seen on the collector side:
```
otel-collector              | InstrumentationLibraryMetrics #0
otel-collector              | InstrumentationLibrary
otel-collector              | Metric #0
otel-collector              | Descriptor:
otel-collector              |      -> Name: swap
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #1
otel-collector              | Descriptor:
otel-collector              |      -> Name: disk
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #2
otel-collector              | Descriptor:
otel-collector              |      -> Name: cpu
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #3
otel-collector              | Descriptor:
otel-collector              |      -> Name: mem
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #4
otel-collector              | Descriptor:
otel-collector              |      -> Name: system
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #5
otel-collector              | Descriptor:
otel-collector              |      -> Name: diskio
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleSum
otel-collector              |      -> IsMonotonic: true
otel-collector              |      -> AggregationTemporality: AGGREGATION_TEMPORALITY_CUMULATIVE
otel-collector              | Metric #6
otel-collector              | Descriptor:
otel-collector              |      -> Name: processes
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
```

Haven't dug into the code to understand what is happening yet though.

codeboten commented on plugins/outputs/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,127 @@
+		return err
+	}
+
+	grpcClientConn, err := grpc.Dial(o.ServiceAddress, grpc.WithInsecure())
```
Using `WithInsecure` here would prevent anyone from using a secure connection, would it be possible to add support for configuring TLS as well? Something like this would be great: https://github.com/influxdata/telegraf/pull/9109/files#diff-9286694f607c96eb8bc3d8f4e1c493a9931f543b009ef2538a03c63dd680418cR1268-R1273

ssoroka replied:
agreed. see our `tls.ClientConfig`

codeboten commented on plugins/outputs/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,127 @@
+  ## Select a schema for metrics: prometheus-v1 or prometheus-v2
+  ## For more information about the alternatives, read the Prometheus input
+  ## plugin notes.
+  # metrics_schema = "prometheus-v1"
```
Some additional configuration options which are supported by the OTLP exporter and would prevent our users from using this plugin if they were not present:
```
#   # Additional resource attributes
#   [outputs.opentelemetry.attributes]
#       "service.name" = "demo"
#
#   # Additional grpc metadata
#   [outputs.opentelemetry.headers]
#       key1 = "value1"
```

See: https://github.com/influxdata/telegraf/pull/9109/files#diff-9286694f607c96eb8bc3d8f4e1c493a9931f543b009ef2538a03c63dd680418cR1275-R1281

Not sure, but maybe if there's another plugin in telegraf that could be used to inject arbitrary fields into incoming data before it arrives to the output plugin that could achieve at least the attributes portion of this configuration?

jacobmarble replied:
Borrowed from your PR.

jacobmarble commented:
> I ran a manual test using telegraf with this plugin to export data to an opentelemetry collector locally and didn't see any data being generated.

I'll work on tests to reproduce and prevent future regression.

jacobmarble commented:
> @jacobmarble this output plugins looks good, but the library it is using is marked experimental. Do you think this plugin is ready for end users and be planned to ship with the next Telegraf feature release?

Fair point. I have added integration tests to the library, removed "experimental" markers, and released v0.1.0. That exposed one bug and a fix (testing FTW!).

jacobmarble commented:
@codeboten I wrote two sets of tests:
- https://github.com/influxdata/influxdb-observability/tree/main/tests-integration
- https://github.com/influxdata/influxdb-observability/tree/main/tests-synthetic

All of the above tests work as intended. Your error "gauge field not found" is a schema issue. I suspect you are consuming metrics with a line protocol schema that is incompatible with your chosen `metrics_schema = "prometheus-v1"` or `metrics_schema = "prometheus-v2"`.

If you are using a schema different to these two, then we need to agree on a strategy for converting other schemata. Not a hard thing, but could we punt it to follow-on issues and PRs against this plugin?

jacobmarble commented:
@ssoroka @codeboten PTAL

ssoroka reviewed (commented):
changes look fine. are there supposed to be test files now?

jacobmarble commented:
> > All of the above tests work as intended. Your error "gauge field not found" is a schema issue. I suspect you are consuming metrics with a line protocol schema that is incompatible with your chosen `metrics_schema = "prometheus-v1"` or `metrics_schema = "prometheus-v2"`.
> 
> should this be possible? shouldn't all line protocol be convertible to valid OTEL ?

I agree, any valid line protocol should find its way into some kind of Otel struct. In the example shared by @codeboten , `influx2otel` has failed to detect the metric type (gauge, sum/counter, histogram) and so has fallen back on gauge. Then it assumes that a field `gauge` exists. Obviously this is incorrect.

I would like to punt this to https://github.com/influxdata/influxdb-observability/issues/2

jacobmarble commented:
> changes look fine. are there supposed to be test files now?

I have added various tests to the library responsible for conversion.
https://github.com/influxdata/influxdb-observability/tree/main/tests-integration
https://github.com/influxdata/influxdb-observability/tree/main/tests-synthetic

This plugin adapts the conversion library for Telegraf, so there isn't a lot to test. Share your thoughts?

codeboten commented:
> All of the above tests work as intended. Your error "gauge field not found" is a schema issue. I suspect you are consuming metrics with a line protocol schema that is incompatible with your chosen `metrics_schema = "prometheus-v1"` or `metrics_schema = "prometheus-v2"`.
> 
> If you are using a schema different to these two, then we need to agree on a strategy for converting other schemata. Not a hard thing, but could we punt it to follow-on issues and PRs against this plugin?

I tested locally again with both schema options and both are giving me errors. With `prometheus-v1` i see the following:
```
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: value type conflict for metric 'swap'; expected 'gauge' or 'untyped', got 'sum
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
2021-06-21T22:57:31Z W! [outputs.opentelemetry] failed to add point: gauge field not found
```

With `prometheus-v2` i see the following:
```
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'swap'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'swap'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'disk'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'mem'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'cpu'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'system'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'system'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'system'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'diskio'
2021-06-21T23:00:31Z W! [outputs.opentelemetry] failed to add point: unexpected measurement name 'processes'
```

This is testing with a default configuration generated by running `./telegraf config > config.toml` and enabling the OpenTelemetry output plugin. The values for the data doesn't appear in the opentelemetry collector logs:

```
otel-collector              | 2021-06-21T22:57:31.504Z	DEBUG	loggingexporter/logging_exporter.go:402	ResourceMetrics #0
otel-collector              | InstrumentationLibraryMetrics #0
otel-collector              | InstrumentationLibrary
otel-collector              | Metric #0
otel-collector              | Descriptor:
otel-collector              |      -> Name: swap
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #1
otel-collector              | Descriptor:
otel-collector              |      -> Name: mem
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #2
otel-collector              | Descriptor:
otel-collector              |      -> Name: disk
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #3
otel-collector              | Descriptor:
otel-collector              |      -> Name: cpu
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #4
otel-collector              | Descriptor:
otel-collector              |      -> Name: system
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleGauge
otel-collector              | Metric #5
otel-collector              | Descriptor:
otel-collector              |      -> Name: diskio
otel-collector              |      -> Description:
otel-collector              |      -> Unit:
otel-collector              |      -> DataType: DoubleSum
otel-collector              |      -> IsMonotonic: true
otel-collector              |      -> AggregationTemporality: AGGREGATION_TEMPORALITY_CUMULATIVE
```

codeboten commented:
I should add that I recompiled telegraf using the following commit: `b5b71662b4df09e3965dbd6129b4a2fb6ce0f324`

jacobmarble commented:
This PR is taking a long time. I'm sorry about that.

I have reproduced the bug that @codeboten has shared. It's clear that requiring a Telegraf/Prometheus schema in this plugin is unreasonable. To properly fix this issue, I'll do a refactor to `influxdb-observability` such that the schema will be inferred from the data, rather than a config.

jacobmarble commented:
@codeboten PTAL

Turns out that many input plugins type their output (gauge, counter, etc), so that needed to be accounted for. The `swap` input plugin gave concise examples to work with.

ssoroka commented:
> I agree, any valid line protocol should find its way into some kind of Otel struct. In the example shared by @codeboten , `influx2otel` has failed to detect the metric type (gauge, sum/counter, histogram) and so has fallen back on gauge. Then it assumes that a field `gauge` exists. Obviously this is incorrect.
> 
> I would like to punt this to [influxdata/influxdb-observability#2](https://github.com/influxdata/influxdb-observability/issues/2)

We're pretty bound by backwards-compatibility; which can make it hard to change later without breaking things. Is there a better option than gauge by default? Is this a case of missing configuration?

jacobmarble commented:
@ssoroka there is no "untyped" metric in OpenTelemetry, unfortunately.

jacobmarble commented:
The failing tests appear to be flakey. Let me know if I'm wrong about that.

codeboten commented on plugins/outputs/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,182 @@
+	tls.ClientConfig
+	Timeout     config.Duration   `toml:"timeout"`
+	Compression string            `toml:"compression"`
+	Headers     map[string]string `toml:"headers"`
```
I'm not seeing where this is currently used, these headers should be setting grpc metadata in the client connections. See https://github.com/influxdata/telegraf/pull/9109/files#diff-2702903f71c7896e7e1d358e50075b5dc48150603438ee5eaea24948b8230d18R111

codeboten reviewed (commented):
I've confirmed that metrics are now being generated, thanks for addressing that. Only question i have remaining is regarding the `headers` configuration option.

ssoroka commented:
Merged! thanks @jacobmarble. @codeboten let's fix that in a separate PR

codeboten commented:
@ssoroka 👍, followed up with the fix in this PR: https://github.com/influxdata/telegraf/pull/9587

shruthikudlur commented:
Is there any plugin that converts Influx Line Protocol Metrix to OTEL Traces.

sjwang90 commented:
@shruthikudlur There is the [OpenTelemetry output plugin](https://github.com/influxdata/telegraf/tree/master/plugins/outputs/opentelemetry)
